### DISCUSSION TITLE
3.0.0-beta09

### DISCUSSION THREAD

**Discussion by Him188** [open-ani/animeko#137] [Announcements]
- 支持数据源代理与设置, 可在 "我的" 的右上角进入
- 数据源选择器: 支持点击已选择的筛选来取消筛选
- 修复有保存的不在列表中的字幕组时, 仍然会用该设置筛选的问题
- 修复 Mikan 资源的时间
- 修复竖屏模式下视频有位移问题

<img width="300" src="https://github.com/Him188/ani/assets/12100985/2b26aba4-1490-4c6f-b610-095383d0e72a" alt="image"/>

----

**附件:**

- `.zip` 是包含 JRE 的解压即用的绿色版本；
- `.dmg`、`.deb`、`.rpm` 和 `.msi` 分别是 macOS、Debian、RedHat 和 Windows 平台的安装程序；
- `.exe` 是包含 JRE 的 Windows 程序，可以直接运行。
- `.jar` 是不包含 JRE 的最小程序，需要配合 JRE 11 及以上使用；
- `.apk` 是 Android 应用，支持的最低 Android 版本为 8.0；

注意，桌面平台仅支持 64 位操作系统。在使用 macOS 的安装程序时可能提示文件损坏，这是因为此版本应用没有签名，可以忽略错误继续使用。

也可扫描以下二维码下载 Android 版本：
![](https://github.com/Him188/ani/releases/download/v3.0.0-beta09/ani-3.0.0-beta09.apk.qrcode.png)

<hr /><em>This discussion was created from the release <a href='https://github.com/Him188/ani/releases/tag/v3.0.0-beta09'>3.0.0-beta09</a>.</em>

**Comment by NeKoOuO** [comment]
怎麼改成組織了啊👀

**Reply by Him188** [reply]
方便拉人进来干活👀

**Reply by Him188** [reply]
找了个朋友来整弹幕服务器

**Reply by NeKoOuO** [reply]
> 找了个朋友来整弹幕服务器

是打算加什麼 Feature 👀

**Reply by Him188** [reply]
发送弹幕, 会发送到我们的服务器
获取时同时从 ani 服务器和弹弹 play 获取

**Comment by nt1r** [comment]
> 找了个朋友来整弹幕服务器

既然要整服务器的话，要不要考虑也整个下载视频用的服务器呢？

我昨晚下载了Android的App体验了一下，谈谈我的初体验吧。目前我觉得最大的痛点在于客户端需要等到视频完全从bt上下载完成之后才能观看视频，期间等待的时间太长了。我都尽量挑选大小在100-200M左右的bt下载了，等待时间还是很长。

有时候用户可能只是想看某一集中的一个时间片段体验体验，但是要等到一整集都下载完成才能观看，Orz

我想象中的解决方案：搞个公共的视频服务器肯定是不可取的（版权、服务器带宽和硬盘所需费用等等都是大问题），但如果是用户自己在本地PC上搭建私有的视频服务器呢？早上先把想看的bt视频下载到私有服务器上，下班后回家直接用App观看可行吗？

我之前试过用Ktor Server搭建视频服务器，Ktor Server支持流媒体文件的分段请求 [Doc](https://ktor.io/docs/server-partial-content.html)，而且对开发者是无感的。App上的ExoPlayer设置好远端资源Url之后，想看第几分钟可以直接拖动进度条，和在线视频网站的体验很类似。

**Reply by Him188** [reply]
非常好的问题, 我两天没发版就是正在整这个

正在改架构支持任意包括本地数据源. 
计划增加自动缓存系统, 用户设置倾向的字幕组和语言等配置后, 可自动缓存下一集

播放器也可以很简单支持 HTTP 视频播放了

缓存支持使用本地文件系统, 也可以实现接入任意私有服务器. 当然确实也可以做一个配套的缓存服务器

**Reply by Him188** [reply]
这个问题和弹幕发送是目前就是最高优先级做的事情了

**Reply by Him188** [reply]
我比较支持你提议的私有服务器方案. 

目前详细计划解决方案如下:

我了解到有朋友自己已经有这样的服务器了 (好像是 SMB), 有脚本爬他追的番到服务器上, 这种情况就可以增加一个 SMB 数据源 (数据源是用番剧标题+剧集序号去搜索资源的地方)
搜索资源时会按数据源的类型区分, 当可以在本地数据源找到资源时, 高优显示/使用

对于 bt 网络的数据, 需要进行 (手动或者自动) 缓存, 可以选择缓存存储在本地或者 SMB. 我有预留对私有缓存服务器支持的可能, 也可以在本地 BT 下载然后传输到远程文件系统. 可以在 app 内关联 bt 来源网站. 作为缓存, 在选择数据源也是高优的

**Reply by Him188** [reply]
目前其实是有变下边播的. 

bt 是分片的, 一个视频文件有 1000 个片这样. 播视频需要读取头尾两三个片取得元数据才能播放. 我会先请求这些元数据片段, 但是可能运气不好会请求到一个比较慢的 peer 上, 就要等很久才能拿到元数据

一个非常自然的想法就是可以同时向 10 个 peer 请求元数据片段, 取最快的, 但是 libtorrent 并没有支持这样的功能... 而我自行去实现 torrent 协议当然是不太现实的

**Reply by Him188** [reply]
私有服务器确实还可以解决种子资源共享的问题

ani 不想成为一个吸血客户端, 做种功能也是非常重要的. 目前安卓端其实我还没测试过这个功能, 但据我了解 libtorrent 应该是默认支持的. 但手机上做种会有电量问题, 用户大概率不会把 ani 挂后台做种

在做缓存服务器的时候需要格外考虑这个, 做一个管理页面之类的 (当然它的优先级没缓存本身高, 早期可以只默认开启限速 1M/s 的上传之类的)

**Reply by nt1r** [reply]
> 非常好的问题, 我两天没发版就是正在整这个
> 

大佬的回复我都看了，我就整体回复一下吧。字里行间能感觉到你对ani的热爱和用心，我随手写下的建议竟然被这么认真的对待，说真的，我工作的上司都没这么热情，我非常感谢你码字详细为我解释付出的时间。

从技术上说，私有服务器做视频缓存+做种的方案应该是可行的。但我想到一个技术之外的问题。昨晚是我第一次使用banngumi.tv，注册账号的时候，bgm着重强调它不提供影视资源。我猜这可能是因为它不想卷入版权纠纷才这么做的。

做种从某种意义上来讲，算得上某种形式的提供盗版资源的行为。如果未来真的因为ani的开源代码被版权诉讼放明确为侵权，又该怎么办。

曲线救国的方法我目前只想到一种，参考破解switch游戏卡带工具的做法，工具本身不提供游戏卡带，由玩家自行购买自愿使用破解工具。类似的，如果私有缓存服务器要做种，可以考虑把它和开源的ani分离开。私有缓存服务器换一个项目地址，或者干脆不开源，让有需求的观众自行寻找下载。

**Reply by Him188** [reply]
确实有这个问题，其实数据源对接部分我也有考虑给它整出去，不太清楚使用其他网站的数据会不会造成版权问题，感觉比较有可能会

**Reply by Him188** [reply]
ani 未来*转型*为大家购买BD之后的弹幕播放器就好了，用户可以自行实现数据源对接*他们的服务器*

*而我还真买了不少BD*
